jni/include/native_engines_stream_support.h

```diff
@@ -119,6 +120,105 @@ class NativeEngineIndexInputMediator {
+                                  JNIEnv *_env,
+                                  jobject _indexOutput)
+      : jni_interface(_jni_interface),
+        env(_env),
+        indexOutput(_indexOutput),
+        bufferArray((jbyteArray) (_jni_interface->GetObjectField(_env,
+                                                                 _indexOutput,
+                                                                 getBufferFieldId(_jni_interface, _env)))),
+        writeBytesMethod(getWriteBytesMethod(_jni_interface, _env)),
+        bufferLength(jni_interface->GetJavaBytesArrayLength(env, bufferArray)),
+        nextWriteIndex() {
+  }
+
+  void writeBytes(uint8_t *source, size_t nbytes) {
+    auto left = nbytes;
+    while (left > 0) {
+      const auto writeBytes = std::min(bufferLength - nextWriteIndex, left);
+
+      // === Critical Section Start ===
+
+      // Get primitive array pointer, no copy is happening in OpenJDK.
+      auto primitiveArray =
+          (jbyte *) jni_interface->GetPrimitiveArrayCritical(env, bufferArray, nullptr);
+
+      // Copy Java bytes to C++ destination address.
+      std::memcpy(primitiveArray + nextWriteIndex, source, writeBytes);
+
+      // Release the acquired primitive array pointer.
+      // JNI_COMMIT tells JVM to copy back the content, but do not free the elems pointer which is pointing to the .
+      // original buffer. However, most OpenJDK just return its underlying buffer pointer rather than creating a
+      // intermediate buffer.
+      jni_interface->ReleasePrimitiveArrayCritical(env, bufferArray, primitiveArray, JNI_COMMIT);
+
+      // === Critical Section End ===
+
+      nextWriteIndex += writeBytes;
+      if (nextWriteIndex >= bufferLength) {
+        callWriteBytesInIndexOutput();
+      }
+
+      source += writeBytes;
+      left -= writeBytes;
+    }  // End while
+  }
+
+  void flush() {
+    if (nextWriteIndex > 0) {
+      callWriteBytesInIndexOutput();
+    }
+  }
+
+ private:
+  static jclass getIndexOutputWithBufferClass(JNIUtilInterface *jni_interface, JNIEnv *env) {
+    static jclass INDEX_OUTPUT_WITH_BUFFER_CLASS =
+        jni_interface->FindClassFromJNIEnv(env, "org/opensearch/knn/index/store/IndexOutputWithBuffer");
+    return INDEX_OUTPUT_WITH_BUFFER_CLASS;
+  }
+
+  static jmethodID getWriteBytesMethod(JNIUtilInterface *jni_interface, JNIEnv *env) {
+    static jmethodID WRITE_METHOD_ID =
+        jni_interface->GetMethodID(env, getIndexOutputWithBufferClass(jni_interface, env), "writeBytes", "(I)V");
+    return WRITE_METHOD_ID;
+  }
+
+  static jfieldID getBufferFieldId(JNIUtilInterface *jni_interface, JNIEnv *env) {
+    static jfieldID BUFFER_FIELD_ID =
+        jni_interface->GetFieldID(env, getIndexOutputWithBufferClass(jni_interface, env), "buffer", "[B");
+    return BUFFER_FIELD_ID;
+  }
+
+  void callWriteBytesInIndexOutput() {
+    auto jclazz = getIndexOutputWithBufferClass(jni_interface, env);
+    jvalue args {.i = nextWriteIndex};
```

Comment:
what is this line doing? my question is mainly on the syntax.

Reply:
will add a comment in that line.
It's union initialization - https://en.cppreference.com/w/c/language/struct_initialization.
Since `writeBytes` has an integer arguments, I'm initializing jvalue as an integer, which `i` represents.